crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -535,43 +551,32 @@ fn execute_move_upgrade<S: StorageView, Mode: ExecutionMode>(
     // Check that this package ID points to a package and get the package we're upgrading.
     let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
 
-    // Run the move + sui verifier on the modules and publish them into the cache.
-    // NB: this will substitute in the original package id for the `self` address in all of these modules.
-    let upgraded_package_modules = publish_and_verify_upgraded_modules::<_, Mode>(
-        context,
-        &module_bytes,
-        current_package.original_package_id(),
-    )?;
+    let mut modules = deserialize_modules::<_, Mode>(context, &module_bytes)?;
+    let runtime_id = current_package.original_package_id();
+    substitute_package_id(&mut modules, runtime_id)?;
 
-    // Full backwards compatibility except that we allow friend function signatures to change.
-    check_compatibility(
-        context,
-        &current_package,
-        &upgraded_package_modules,
-        upgrade_ticket.policy,
-    )?;
+    // Upgraded packages share their predecessor's runtime ID but get a new storage ID.
+    let storage_id = context.tx_context.fresh_id();
 
-    // Read the package dependencies.
-    let dependency_packages = fetch_packages(context, &dep_ids)?;
+    let dependencies = fetch_packages(context, &dep_ids)?;
+    let package = context.upgrade_package(storage_id, &current_package, &modules, &dependencies)?;
```

Comment:
I went kind of the opposite route in https://github.com/MystenLabs/sui/pull/9567 - create linkage table (and type origin table) here and pass it on directly to `MovePackage` creation (which also involved move_package module refactoring). I like the approach in this PR, though! 

I am not sure what's the current idea for setting linkage context is, but in https://github.com/MystenLabs/sui/pull/9567 I had two ways: "compute" (where we give the constructor just the `ObjectID` and have it retrieve all the needed info by loading a given package object) and "set" (which would be used here, where we explicitly pass the tables to the constructor).

I think we may either still need this distinction (as the "compute" would not work here due to the package not being available in storage yet) or somehow incorporate the list of packages published in the current programmable transaction into the "compute" version.

I hope this comment makes at least some sense... It's nothing that should prevent landing this, but rather something to keep in mind.

Reply:
Yes indeed, we still need both cases, because in most cases we won't have the `MovePackage` already to hand!